python/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+// the proposed implementation is exposing static method returning an
+// initialized instance of DriverInitParam struct
+/// Manages additional driver initialization parameters
+#[pyclass]
+#[derive(Debug, Clone)]
+struct DriverInitParam {
+    pub param_type: DriverInitParamType,
+    pub param_data_string: String,
+}
+
+#[pymethods]
+impl DriverInitParam {
+    /// initialize a DriverInitParam for the KVM driver, with a Unix socket
+    #[staticmethod]
+    fn kvmi_unix_socket(socket: &str) -> Self {
+        DriverInitParam {
+            param_type: DriverInitParamType::KVMiUnixSocket,
+            param_data_string: socket.to_string(),
+        }
+    }
+}
+
+/// Main class to interact with libmicrovmi
+// A class marked as unsendable will panic when accessed by another thread.
+// TODO: make Introspectable trait inherit Send trait, and make the drivers implementation
+// compatible
+#[pyclass(unsendable)]
+struct Microvmi {
+    driver: Box<dyn rapi::Introspectable>,
+}
+
+#[pymethods]
+impl Microvmi {
+    // enums are not available in PyO3 (yet)
+    // TODO: docstring is not exposed in Python (bug ?)
+    /// initializes libmicrovmi from the specified domain name
+    /// if driver_type is None, every driver compiled in libmicrovmi will be tested,
+    /// and the first one that succeeds will be returned, or an error
+    ///
+    /// Args:
+    ///     domain_name (str): the domain name
+    ///     driver_type (int, optional): the hypervisor driver type on which the library should be initialized.
+    ///     init_param (DriverInitParam, optional): additional initialization parameters for driver initialization
+    #[new]
+    #[args(domain_name, driver_type = "None", init_param = "None")]
+    fn new(
+        domain_name: &str,
+        driver_type: Option<u32>,
+        init_param: Option<DriverInitParam>,
+    ) -> PyResult<Self> {
+        info!("Microvmi Python init");
+        // convert Python DriverType to rust API DriverType
+        debug!(
+            "Microvmi Python init driver_type: {:?}, init_param: {:?}",
+            driver_type, init_param
+        );
+        let rust_driver_type: Option<rapi::DriverType> = if let Some(drv_type) = driver_type {
+            Some(match drv_type {
+                DriverType::HYPERV => Ok(rapi::DriverType::HyperV),
+                DriverType::KVM => Ok(rapi::DriverType::KVM),
+                DriverType::VIRTUALBOX => Ok(rapi::DriverType::VirtualBox),
+                DriverType::XEN => Ok(rapi::DriverType::Xen),
+                _ => Err(PyValueError::new_err(format!(
+                    "Invalid value for DriverType: {}",
+                    drv_type
+                ))),
+            }?)
+        } else {
+            None
+        };
+        // convert Python DriverInitParam to rust API DriverinitParam
+        let rust_init_param: Option<rapi::DriverInitParam> = if let Some(param) = init_param {
+            Some(
+                #[allow(unreachable_patterns)]
```

Comment:
Why this?

Reply:
because clippy would complain that since there was only one pattern match, it could be replaced by the matched body itself

Reply:
Shouldn't you make `DriverInitParam` non_exhaustive instead?

Reply:
non exhaustive is more dangerous.
it means that when we add a new enum, this match arm will still compile, even though we forgot to handle the new enum value, so it's not the right solution here

Reply:
Right, but this also means that adding values to `DriverInitParam` (I mean the one we expose to our users) is always a breaking change. But if that's okay with you we can leave it like this.

Reply:
SInce the python library recompiles libmicrovmi and distributes it, this is not an issue.